### ISSUE TITLE
[Feature]: return `custom_llm_provider` in litellm completion call

### ISSUE THREAD

**Issue by ishaan-jaff** [BerriAI/litellm#1409]
### The Feature

extract which API provider was used when litellm.completion is called with a list of fallbacks

Sorry for the late reply, I'd be happy to. 

Here's a minimal [contrived] example of how we're using litellm.

Class MyInferenceClass:
    # A list of models litellm will fallback to (NOTIONAL)
    MODELS = ['openai/gpt-35', 'auzre/gpt-35`, 'anthropic/claude']

    def do_inference(self, messages):
      response = litellm.completion(
          # Use the first model as the primary model.
          model=self.MODELS[0],

          messages=messages,

          # Provide litellm with our list of fallback models.
          fallbacks=self.MODELS,
      )

Note that MyInferenceClass.MODELS is how we're defining our fallback models, which may be from different API providers. 

Right now, response is a litellm.ModelResponse (https://github.com/BerriAI/litellm/blob/83943151734f0f8f19771a473f94730edfd598c4/litellm/utils.py#L351)

litellm.ModelResponse does have a model attribute, but I'm noticing that it seems to only contain the piece of the model string after the provider.

For example, if openai/gpt-35 was used to generate  response: ModelResponse = litellm.completion(...),  then response.model == gpt-35. I would like to be able to do something like either:

response.model == openai/gpt-35
or
response.model_provider == openai

### Motivation, pitch

user request

**Comment by elsell** [comment]
@ishaan-jaff I'm the OP from Discord. Thanks for tracking this here.

I've polished up the formatting if you'd like to update the original issue 😄 

<br/><br/><br/>

## The Feature
Include API _provider_ in the `ModelResponse` object, not just the model name.

## Motivation, pitch

**TLDR: Being able to track which models _and_ API providers are being used in a production environment when using fallback models is invaluable to assess system health and provide visibility into the fallback mechanism.**

Here's an example, showing what happens now, and what I'd like to happen:


#### How it works now

```python
Class MyInferenceClass:
    # A list of models litellm will fallback to (NOTIONAL)
    MODELS = ['openai/gpt-35', 'auzre/gpt-35', 'anthropic/claude']

    def do_inference(self, messages):
      response = litellm.completion(
          # Use the first model as the primary model.
          model=self.MODELS[0],

          messages=messages,

          # Provide litellm with our list of fallback models.
          fallbacks=self.MODELS,
      )
```

> [!Note]
> `MyInferenceClass.MODELS` is how we're defining our fallback models, which may be from different API providers. 


Right now, the `response` variable is a `litellm.ModelResponse`: https://github.com/BerriAI/litellm/blob/83943151734f0f8f19771a473f94730edfd598c4/litellm/utils.py#L351


`litellm.ModelResponse` does have a `model` attribute, but I'm noticing that it seems to only contain the piece of the model string after the provider.

#### How I'd like it to work
For example, if `openai/gpt-35` was used to generate  
```python
response: ModelResponse = litellm.completion(...)
```
then `response.model == 'gpt-35'`. I would like to be able to do something like either:

`response.model == 'openai/gpt-35'`
or
`response.model_provider == 'openai'`

**Comment by ghost** [comment]
Hey @elsell any reason you're not just using the Router - https://docs.litellm.ai/docs/routing

This is already tracked there via `previous_models` in the logging metadata

**Comment by ghost** [comment]
Is this what you wanted? 
https://github.com/BerriAI/litellm/blob/43533812a7a3872b79b1183d2db259441b6494a7/litellm/router.py#L1189

**Comment by ishaan-jaff** [comment]
## Thinking through this:
- We can add the provider in the litellm response (this is not ideal since it would be a breaking change for litellm users in prod) 
- We can add a hidden param,` _litellm_model = openai/gpt-35` 
- We can expose this information in litellm callbacks https://docs.litellm.ai/docs/observability/custom_callback (I think we already do this) 

What's ideal for you @elsell ?

**Comment by ishaan-jaff** [comment]
@elsell You can already do this with custom callbacks on litellm. Let me know what the ideal experience for you would be  
```python
def custom_callback(
    kwargs,                 # kwargs to completion
    completion_response,    # response from completion
    start_time, end_time    # start/end time
):
    # Your custom code here
    try:
        print("kwargs", kwargs)
        litellm_model_name = kwargs.get("model") + "/" + kwargs.get("litellm_params").get("custom_llm_provider")

        print("litellm_model_name=", litellm_model_name)
    except Exception as e:
        print(e)
        raise e

def test_completion_mistral_api():
    try:
        litellm.success_callback=[custom_callback]
        # litellm.set_verbose = True
        response = completion(
            model="mistral/mistral-tiny",
            messages=[
                {
                    "role": "user",
                    "content": "Hey, how's it going?",
                }
            ],
        )
        # Add any assertions here to check the response
        print("response_ms=", response._response_ms)
    except Exception as e:
        pytest.fail(f"Error occurred: {e}")


test_completion_mistral_api()
```

**Comment by elsell** [comment]
> Hey @elsell any reason you're not just using the Router - https://docs.litellm.ai/docs/routing
> 
> This is already tracked there via `previous_models` in the logging metadata

@krrishdholakia Thank you for pointing me toward `Router`, that is a helpful object that I think I will use to handle more complex fallback logic in my application.

However, I don't think that `Router.previous_models` is ideal since it only tells me the models that failed to provide a completion (correct me if I'm wrong.) It's great for tracking errors, though.

What I'm looking for is an easy way to match a completion with the _provider_ that was used to make that completion.

**Comment by elsell** [comment]
> ## Thinking through this:
> 
>     * We can add the provider in the litellm response (this is not ideal since it would be a breaking change for litellm users in prod)
> 
>     * We can add a hidden param,` _litellm_model = openai/gpt-35`
> 
>     * We can expose this information in litellm callbacks https://docs.litellm.ai/docs/observability/custom_callback (I think we already do this)
> 
> 
> What's ideal for you @elsell ?

@ishaan-jaff I understand that I'm biased toward my use case, so please push back if you think this way forward isn't good for the community at large.

For me, a custom callback is cool and powerful, but it just adds unneeded complexity to my system. 

I prefer being able to access the provider via the `ModelResponse` (after all, it seems to make sense for it to be there semantically, I think.)

Obviously, it doesn't make sense to update the behavior of `ModelResponse.model` due to breaking existing code, so I think a hidden param as you mentioned is a reasonable solution.

However, do you think it would make more sense to add a public attribute called something like `ModelResponse.custom_llm_provider`? I'm fine with either solution.

**Comment by ghost** [comment]
@elsell that's a good idea. 

We can add it to `_hidden_params` so it doesn't break the openai object format 

You can then access it via: `ModelResponse._hidden_params.custom_llm_provider`

**Comment by ishaan-jaff** [comment]
PR here: https://github.com/BerriAI/litellm/pull/1432

**Comment by ishaan-jaff** [comment]
okay @elsell we support getting the custom_llm_provider like this now - is this what you were looking for ? 

```python
response = completion(
    model="azure/chatgpt-v-2",
    messages=messages,
    api_key=old_key,
    max_tokens=10,
)
print(f"response: {response}")

print("Hidden Params", response._hidden_params)
assert response._hidden_params["custom_llm_provider"] == "azure"
```

**Comment by ghost** [comment]
@ishaan-jaff Does this work with streaming?

**Comment by elsell** [comment]
> okay @elsell we support getting the custom_llm_provider like this now - is this what you were looking for ?
> 
> ```python
> response = completion(
>     model="azure/chatgpt-v-2",
>     messages=messages,
>     api_key=old_key,
>     max_tokens=10,
> )
> print(f"response: {response}")
> 
> print("Hidden Params", response._hidden_params)
> assert response._hidden_params["custom_llm_provider"] == "azure"
> ```
@ishaan-jaff Yes, that looks great!

Although I am curious to know the answer to @krrishdholakia's question - more specifically, would the custom provide be included in the ModelResponse object created by `litellm.stream_chunk_builder`?

**Comment by ishaan-jaff** [comment]
We return this in the stream chunks - will add testing to ensure `stream_chunk_builder` returns this before merging https://github.com/BerriAI/litellm/pull/1432/files

**Comment by elsell** [comment]
> We return this in the stream chunks - will add testing to ensure `stream_chunk_builder` returns this before merging https://github.com/BerriAI/litellm/pull/1432/files

Love it. Thank you so much!

**Comment by ishaan-jaff** [comment]
@elsell

Do you use `litellm.stream_chunk_builder` in callbacks ? . Want to make sure this works well for your use case - can you let me know how / when you use `litellm.stream_chunk_builder`

**Comment by elsell** [comment]
> @elsell
> 
> Do you use `litellm.stream_chunk_builder` in callbacks ? . Want to make sure this works well for your use case - can you let me know how / when you use `litellm.stream_chunk_builder`

@ishaan-jaff 
I am not using `stream_chunk_builder` in callbacks, I'm using it to assemble all chunks after a generation. 

As pseudo code,

```python

chunks = []

for chunk in do_litellm_streaming_completion():
    chunks.append(chunk)

complete_response = litellm.stream_chunk_builder(
    chunks=chunks, messages=messages
)
```

**Comment by ishaan-jaff** [comment]
merged here, https://github.com/BerriAI/litellm/commit/7fc8ccf854499b52f4689ae9364384355a07cb48 

works with the way you're using it

**Comment by ishaan-jaff** [comment]
@elsell in prod: https://github.com/BerriAI/litellm/releases/tag/v1.17.5 

closing this issue - feel free to re-open if it's not solved well 

Let me know if you have any feedback on how it can be improved
